src/bvals/comms/coalesced_buffers.hpp

```diff
@@ -0,0 +1,207 @@
+//========================================================================================
+// (C) (or copyright) 2024. Triad National Security, LLC. All rights reserved.
+//
+// This program was produced under U.S. Government contract 89233218CNA000001 for Los
+// Alamos National Laboratory (LANL), which is operated by Triad National Security, LLC
+// for the U.S. Department of Energy/National Nuclear Security Administration. All rights
+// in the program are reserved by Triad National Security, LLC, and the U.S. Department
+// of Energy/National Nuclear Security Administration. The Government is granted for
+// itself and others acting on its behalf a nonexclusive, paid-up, irrevocable worldwide
+// license in this material to reproduce, prepare derivative works, distribute copies to
+// the public, perform publicly and display publicly, and to permit others to do so.
+//========================================================================================
+
+#ifndef BVALS_COMMS_COALESCED_BUFFERS_HPP_
+#define BVALS_COMMS_COALESCED_BUFFERS_HPP_
+
+#include <map>
+#include <memory>
+#include <set>
+#include <string>
+#include <unordered_map>
+#include <utility>
+#include <vector>
+
+#include "basic_types.hpp"
+#include "bvals/comms/bnd_id.hpp"
+#include "bvals/comms/bvals_utils.hpp"
+#include "bvals/neighbor_block.hpp"
+#include "coordinates/coordinates.hpp"
+#include "interface/variable.hpp"
+#include "mesh/mesh.hpp"
+#include "mesh/meshblock.hpp"
+#include "utils/communication_buffer.hpp"
+
+namespace parthenon {
+// Structure containing the information required for sending coalesced
+// messages between ranks
+
+struct uid_set_hash {
+  std::size_t operator()(const std::set<Uid_t> &in) const {
+    std::size_t lhs{0};
+    for (const auto &uid : in) {
+      std::size_t rhs = std::hash<Uid_t>()(uid);
+      lhs ^= rhs + 0x9e3779b9 + (lhs << 6) + (lhs >> 2);
+    }
+    return lhs;
+  }
+};
+
+struct CoalescedBuffer {
+  using buf_t = BufArray1D<Real>;
+
+  // Rank that these buffers communicate with
+  BoundaryType b_type;
+  int other_rank;
+  int partition;
+  mpi_comm_t comm_;
+  Mesh *pmesh;
+  bool sender;
+
+  using var_buf_t = CommBuffer<buf_pool_t<Real>::owner_t>;
+  std::map<Uid_t, std::vector<std::pair<BndId, var_buf_t *>>> coalesced_info_buf;
+  std::set<Uid_t> all_vars;
+  std::unordered_map<std::set<Uid_t>, ParArray1D<BndId>, uid_set_hash> bnd_ids_device_map;
```

Comment:
The ` ParArray1D<BndId>` are view of views and might be related to the hangs observed in Riot.

Reply:
yes, I can confirm that was the source of the initial hang that we were seeing (although once past this one another hang occurs in Riot that I haven't fully debugged, but probably related to using partitions and the partition of everything).